### PR TITLE
enable std::assume_aligned

### PR DESCRIPTION
**Pull request by fsb4000**
Resolves https://github.com/microsoft/STL/issues/47

docs: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1007r3.pdf

### CODE REVIEW

File: stl/inc/yvals_core.h [hunk 1]
```diff
@@ -176,6 +176,7 @@
 // P0966R1 string::reserve() Should Not Shrink
 // P1001R2 execution::unseq
 // P1006R1 constexpr For pointer_traits<T*>::pointer_to()
+// P1007R3 std::assume_aligned
```
Comment (StephanTLavavej) [inline stl/inc/yvals_core.h:179]: Nitpick: We conventionally copy the issue title exactly, so this should follow #47:
```suggestion
// P1007R3 assume_aligned()
```
(The background here is that I "clean up" the various paper titles by simplifying them, making capitalization consistent, removing `std::` qualification, etc. and I like to see that reflected in this list. It doesn't really matter in this case although some papers have very confusing titles that do benefit from this cleanup.)

Comment (StephanTLavavej) [inline stl/inc/yvals_core.h:179]: `yvals_core.h` needs to define the feature-test macro mentioned in #47 (in the C++20 section, in sorted order).

Then, it needs to be tested in `tests/std/tests/VSO_0157762_feature_test_macros/test.cpp`.

File: tests/std/tests/P1007R3_assume_aligned/test.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <cassert>
+#include <memory>
+
+static void can_change_value_via_assume_aligned(int* p) {
+    int* p1 = std::assume_aligned<256>(p);
+    p1[0]   = 10;
+    p1[1]   = -15;
+}
+
+static constexpr int can_read_constexpr_value_via_assume_aligned_and_sum_it(int const* p, size_t size) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P1007R3_assume_aligned/test.cpp:13]: Technically need to include `<stddef.h>` for `size_t`, or `<cstddef>` for `std::size_t` (if you want to qualify the occurrences).

File: tests/std/tests/P1007R3_assume_aligned/test.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <memory>
+
+static void can_change_value_via_assume_aligned(int* p) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P1007R3_assume_aligned/test.cpp:7]: As this is a single-source-file test, `static` doesn't do anything; I recommend removing it from these functions.

File: stl/inc/memory [hunk 4]
```diff
@@ -3079,15 +3079,21 @@ inline void* align(size_t _Bound, size_t _Size, void*& _Ptr, size_t& _Space) noe
...
+#if 1 // TRANSITION: delete after __builtin_assume_aligned will be constexpr
+    if (!std::is_constant_evaluated()) {
+        return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, _Nx));
+    }
 
     return _Ptr;
+#else
+    // this enforces the requirement that _Nx be a power of two
+    return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, _Nx));
+#endif // TRANSITION: delete after __builtin_assume_aligned will be constexpr
```
Comment (StephanTLavavej) [inline stl/inc/memory:3094]: 1. We conventionally use the `_STD` macro in product code, which expands to `::std::`. This theoretically defends against pathological user namespaces. (Even though `is_constant_evaluated()` takes no arguments and is therefore invulnerable to argument-dependent lookup hijacking.)
2. We conventionally prefer to avoid negated tests, unless attempting to do so over-complicates the code. This is because it is generally easier to read and understand positive tests ("if cat", instead of "if not cat"). In this case, I believe that flipping this around to `if (_STD is_constant_evaluated())` guarding `return _Ptr;` would have a bonus simplifying effect - that branch could be marked as `TRANSITION`, and then the following usage of `__builtin_assume_aligned` wouldn't need to be repeated (in active and inactive forms).

File: stl/inc/memory [hunk 5]
```diff
@@ -3079,15 +3079,21 @@ inline void* align(size_t _Bound, size_t _Size, void*& _Ptr, size_t& _Space) noe
     return _Ptr;
 }
 
-#if _HAS_CXX20 && 0
+#if _HAS_CXX20
 template <size_t _Nx, class _Ty>
 _NODISCARD constexpr _Ty* assume_aligned(_Ty* _Ptr) noexcept /* strengthened */ {
-    // this enforces the requirement that _Nx be a power of two
-    __builtin_assume_aligned(_Ptr, _Nx);
+#if 1 // TRANSITION: delete after __builtin_assume_aligned will be constexpr
```
Comment (StephanTLavavej) [inline stl/inc/memory:3085]: What compilers (MSVC, Clang, EDG IntelliSense) don't support `constexpr __builtin_assume_aligned`? We need to file bug reports (on DevCom and LLVM) if we want this to be fixed, then we need to cite the bug numbers in this comment.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks for implementing this feature (over 4% of the remaining features :joy_cat:)! I think it'll be ready to merge after one round of revisions.

File: stl/inc/memory (hunk 5 above)
Comment (CaseyCarter) [reply stl/inc/memory:3085]: Sorry: this was my fault. I suggested the `is_constant_evaluated` hack to unblock this work while I investigate what regressed the `constexpr` intrinsic in MSVC. (It's been so long since the compiler backend team implemented this I started doubting my memory that the intrinsic was working in `constexpr`.) I determined that it definitely _should_ be working but is not - some `constexpr` change has regressed it.

@fsb4000 do you mind filing a DevCom bug noting that neither MSVC nor EDG/IntelliSense accept this intrinsic in `constexpr`? I could take care of it, otherwise.

Comment (fsb4000) [reply stl/inc/memory:3085]: @CaseyCarter I can create issue. I have test case which works for clang-cl and doesn't work for MSVC.
How do you check EDG/IntelliSense?

File: stl/inc/memory (hunk 4 above)
Comment (fsb4000) [reply stl/inc/memory]: I read your explanation about `_STD` here and in discord. And I made the same mistake :(
Sorry you had to repeat this for the thousandth time.

Comment (StephanTLavavej) [reply stl/inc/memory]: Heh, no problem - it takes a long time to get used to our unusual conventions!

File: stl/inc/memory (hunk 5 above)
Comment (StephanTLavavej) [reply stl/inc/memory]: You can take an MSVC command line and add `/c /BE` to say "compile-only (don't link) with the secret undocumented option that activates the EDG front-end".

File: tests/std/tests/P1007R3_assume_aligned/test.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+constexpr void constexpr_assume_aligned_can_read_value() {
+    alignas(256) constexpr int arr[3] = {10, 11, 9};
+    static_assert(can_read_constexpr_value_via_assume_aligned_and_sum_it(arr, 3) == 30);
+}
+
+void assume_aligned_can_read_value() {
+    alignas(256) constexpr int arr[3] = {10, 11, 9};
+    assert(can_read_constexpr_value_via_assume_aligned_and_sum_it(arr, 3) == 30);
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/P1007R3_assume_aligned/test.cpp:38]: There's a technique to avoid repeating logic for runtime and `constexpr` test coverage. Here's what it looks like:

```
C:\Temp>type meow.cpp
```
```cpp
#include <assert.h>
#include <functional>
using namespace std;

constexpr bool test_multiplies() {
    multiplies<> function_object;
    assert(function_object(3, 4) == 12); // note: plain assert
    return true;
}

int main() {
    test_multiplies(); // test runtime behavior
    static_assert(test_multiplies()); // test compiletime behavior
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp

C:\Temp>
```

That is, if you write a `constexpr` test function, it can contain ordinary `assert`s to verify things. If invoked normally at runtime, the `assert` statements will verify runtime behavior. If you `static_assert` the test function, that will force compile-time evaluation, and any failed `assert`s will be a compiler error. (This is not obvious - a successful assert does nothing, so it's constexpr-friendly. A failed assert will try to display a dialog and terminate the program, which is inherently constexpr-forbidden, so it will trigger a compiler error all by itself.)

I will push a simplification for this; I believe it can be extended to the modification test too.

**Review by StephanTLavavej — COMMENTED** [review]
This looks good, thanks! I pushed a couple of changes to the product code:

1. Because Clang works, we should use `#ifndef __clang__` to guard the workaround, so it activates for MSVC and EDG only.
2. We can "extract" the non-workaround code, so it doesn't have to be repeated.

I also changed the `#endif` comment, to make it super clear that the whole block is the workaround that should be removed. (Sometimes when a workaround is necessary for years, it's easy to forget what we should do when the relevant bug is finally fixed.)

I'll push one more change to the test code, then I think this will be ready for final review!

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
One more thing that I forgot to mention - squashing your development history down to a single commit before creating a PR (as you did here) is great. (Creating a PR with separate commits can make sense for mostly-unrelated changes or multi-layered rewrites; I do that occasionally.) However, after a PR has been created, further changes should generally be pushed without rewriting history. This is because GitHub's interface makes it relatively easy to see "commits added since my last review", but makes it very hard to see what happened between force-pushes.

Totally not a problem for this PR, just advice for larger PRs in the future. :smiley_cat:

File: stl/inc/memory [hunk 7]
```diff
@@ -3079,15 +3079,21 @@ inline void* align(size_t _Bound, size_t _Size, void*& _Ptr, size_t& _Space) noe
     return _Ptr;
 }
 
-#if _HAS_CXX20 && 0
+#if _HAS_CXX20
 template <size_t _Nx, class _Ty>
 _NODISCARD constexpr _Ty* assume_aligned(_Ty* _Ptr) noexcept /* strengthened */ {
```
Comment (CaseyCarter) [inline stl/inc/memory:3084]: `_Ptr` could be `const` here, since we don't modify it in the function.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/memory (hunk 7 above)
Comment (CaseyCarter) [reply stl/inc/memory:3084]:
```suggestion
_NODISCARD constexpr _Ty* assume_aligned(_Ty* const _Ptr) noexcept /* strengthened */ {
```

File: stl/inc/memory [hunk 8]
```diff
@@ -3079,15 +3079,21 @@ inline void* align(size_t _Bound, size_t _Size, void*& _Ptr, size_t& _Space) noe
...
+_NODISCARD constexpr _Ty* assume_aligned(_Ty* const _Ptr) noexcept /* strengthened */ {
+#ifndef __clang__ // TRANSITION, DevCom-1190972
+    if (_STD is_constant_evaluated()) {
+        return _Ptr;
+    }
+#endif // ^^^ workaround ^^^
+
     // this enforces the requirement that _Nx be a power of two
-    __builtin_assume_aligned(_Ptr, _Nx);
+    (void) __builtin_assume_aligned(_Ptr, _Nx);
```
Comment (AlexGuteniev) [inline stl/inc/memory:3092]: Shouldn't return resulting poitner cast back to `_Ty` ?

**Review by AlexGuteniev — CHANGES_REQUESTED** [review]

File: stl/inc/memory (hunk 8 above)
Comment (StephanTLavavej) [reply stl/inc/memory:3092]: That's a good question; what type does the builtin return? There was a sentence in the Standard about compilers maybe caring about whether the specific value returned by `assume_aligned` is used; is the builtin's behavior with respect to that documented by anyone?

Comment (AlexGuteniev) [reply stl/inc/memory:3092]: in gcc documentation only, I assume

Comment (AlexGuteniev) [reply stl/inc/memory:3092]: https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

Comment (fsb4000) [reply stl/inc/memory:3092]: cast from 'void *' is not allowed in a constant expression

or you mean:
 if constant expression return original pointer
if ordinal function call return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, _Nx))

Comment (fsb4000) [reply stl/inc/memory:3092]: @StephanTLavavej >what type does the builtin return?

void*

https://imgur.com/a/qmrmIQI

Comment (AlexGuteniev) [reply stl/inc/memory:3092]: I'd want `static_cast<_Ty*>`.
Since it is not clear if the parameter of intrinsic is also assumed to be aligned, I'm for using the result

Comment (CaseyCarter) [reply stl/inc/memory:3092]: MSVC does not care, but I'm uncertain about Clang. Clang documents that this intrinsic is implemented with GCC's semantics and the GCC manual is unclear.

TLDR we should probably use the return value of the intrinsic in non-constant-expressions. Alignment is meaningless in every `constexpr` evaluator I'm aware of so we may as well:
```c++
if (_STD is_constant_evaluated()) {
  return _Ptr;
} else {
  return static_cast<_Ty*>(__builtin_assume_aligned(_Ptr, N));
}
```
and ignore that MSVC and EDG reject calls to the intrinsic in constant expressions.

Comment (fsb4000) [reply stl/inc/memory]: What about https://developercommunity.visualstudio.com/content/problem/1190972/-builtin-assume-aligned-is-not-constexpr.html ?
Should I close it?

**Review by StephanTLavavej — APPROVED** [review]
Thanks! I pushed a small change to restore the comment explaining how the Standard's power-of-two requirement is being enforced.

File: stl/inc/memory (hunk 8 above)
Comment (StephanTLavavej) [reply stl/inc/memory]: Since the `static_cast` is necessary, and `constexpr`-incompatible, it seems reasonable to close the bug report - even if the compiler made the builtin constexpr, it wouldn't help anything.

Comment (fsb4000) [reply stl/inc/memory]: Done, I closed a bug in the https://developercommunity.visualstudio.com

And I filed a bug in gcc bugzilla. (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97132)

https://gcc.godbolt.org/z/KK6W8G

They use static_cast without is_constant_evaluated.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for aligning the STL's feature set with the C++20 Standard, and congratulations on implementing your first feature! :tada: :rocket: :smiley_cat:
